subsys/bluetooth/services/cgms/cgms_racp.c

```diff
@@ -0,0 +1,473 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <zephyr/types.h>
+#include <zephyr/zephyr.h>
+#include <string.h>
+#include <bluetooth/services/cgms.h>
+#include "cgms_internal.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(cgms, CONFIG_BT_CGMS_LOG_LEVEL);
+
+/**@brief Record Access Control Point opcodes. */
+enum cgms_racp_opcode {
+	RACP_OPCODE_RESERVED = 0,
+	RACP_OPCODE_REPORT_RECS = 1,
+	RACP_OPCODE_DELETE_RECS = 2,
+	RACP_OPCODE_ABORT_OPERATION = 3,
+	RACP_OPCODE_REPORT_NUM_RECS = 4,
+	RACP_OPCODE_NUM_RECS_RESPONSE = 5,
+	RACP_OPCODE_RESPONSE_CODE = 6,
+};
+
+/**@brief Record Access Control Point operators. */
+enum cgms_racp_operator {
+	RACP_OPERATOR_NULL = 0,
+	RACP_OPERATOR_ALL = 1,
+	RACP_OPERATOR_LESS_OR_EQUAL = 2,
+	RACP_OPERATOR_GREATER_OR_EQUAL = 3,
+	RACP_OPERATOR_RANGE = 4,
+	RACP_OPERATOR_FIRST = 5,
+	RACP_OPERATOR_LAST = 6,
+	RACP_OPERATOR_RFU_START = 7,
+};
+
+/**@brief Record Access Control Point Operand Filter Type Value. */
+enum cgms_racp_filter {
+	RACP_OPERAND_FILTER_TYPE_TIME_OFFSET = 1,
+	RACP_OPERAND_FILTER_TYPE_FACING_TIME = 2,
+};
+
+/**@brief Record Access Control Point response codes. */
+enum cgms_racp_rsp_code {
+	RACP_RESPONSE_RESERVED = 0,
+	RACP_RESPONSE_SUCCESS = 1,
+	RACP_RESPONSE_OPCODE_UNSUPPORTED = 2,
+	RACP_RESPONSE_INVALID_OPERATOR = 3,
+	RACP_RESPONSE_OPERATOR_UNSUPPORTED = 4,
+	RACP_RESPONSE_INVALID_OPERAND = 5,
+	RACP_RESPONSE_NO_RECORDS_FOUND = 6,
+	RACP_RESPONSE_ABORT_FAILED = 7,
+	RACP_RESPONSE_PROCEDURE_NOT_DONE = 8,
+	RACP_RESPONSE_OPERAND_UNSUPPORTED = 9,
+};
+
+/* The number of measurement that can be stored */
+#define RECORD_NUM       CONFIG_BT_CGMS_NUMBER_MEAS_RECORD_MAX
+
+struct cgms_meas_db {
+	uint16_t count;
+	struct cgms_meas meas[RECORD_NUM];
+	uint16_t head;
+};
```

Comment:
Consider to replace this structure with a data structure that Zephey provides e.g single-linked list.
Using the list, we have an API available for the basic operations that are needed:
- add item - `sys_slist_append()`
- remove item - `sys_slist_remove()`
- iterate on a list  - `SYS_SLIST_FOR_EACH_NODE`

Memory for elements/nodes can be allocated statically. When it is full, the oldest node is removed.
Using the data structure from Zephy, we avoid manual management of elements in such a structure like:
```C
if (database.count < ARRAY_SIZE(database.meas)) {
	memcpy(&database.meas[(database.head + database.count)
		% ARRAY_SIZE(database.meas)],
		&meas, sizeof(meas));
	database.count++;
} else {
	memcpy(&database.meas[(database.head + database.count)
		% ARRAY_SIZE(database.meas)],
		&meas, sizeof(meas));
	database.head = (database.head + 1) % ARRAY_SIZE(database.meas);
}
```
I think it will make the implementation simpler.